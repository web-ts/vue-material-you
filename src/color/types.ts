export type DarkModeState = "auto" | "light" | "dark";
export type DarkModeSystemState = "light" | "dark";

export interface DarkModeSettings {
  userMode: DarkModeState;
  systemMode: DarkModeSystemState;
}
