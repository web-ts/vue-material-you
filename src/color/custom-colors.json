{
  "light": {
    "--md-sys-color-disabled": "31, 31, 31"
  },
  "dark": {
    "--md-sys-color-disabled": "227, 227, 227"
  }
}
