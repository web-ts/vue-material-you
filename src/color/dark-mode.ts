import { DarkModeSettings, DarkModeState } from "./types";

/**
 * The default dark mode settings.
 */
export const settings: DarkModeSettings = reactive({
  userMode: "auto",
  systemMode: window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light"
});

export function onSchemeChange(event: MediaQueryListEvent) {
  const isDark = event.matches;

  settings.systemMode = isDark ? "dark" : "light";
}

window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", onSchemeChange);

export const isDark = computed(() => {
  if (settings.userMode === "auto") return settings.systemMode === "dark";

  return settings.userMode === "dark";
});

export function setDarkMode(mode: DarkModeState) {
  settings.userMode = mode;
}

export const darkMode = computed({
  get: () => settings.userMode,
  set: (mode) => setDarkMode(mode)
});
