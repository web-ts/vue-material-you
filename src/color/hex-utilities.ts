/**
 * Check if the color is a valid hex color
 */
export function isHexColor(color: string) {
  return /^#[0-9A-F]{6}$/i.test(color);
}

/**
 * Convert hex color to "r, g, b" string
 */
export function rgbFromHex(hex: string) {
  const h = hex.trim();
  const r = "0x" + h[1] + h[2];
  const g = "0x" + h[3] + h[4];
  const b = "0x" + h[5] + h[6];

  return `${Number(r)}, ${Number(g)}, ${Number(b)}`;
}
