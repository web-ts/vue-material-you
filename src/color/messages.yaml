---
INVALID_HEX: 'Invalid HEX color. Please use a valid HEX (eg. #12F349) color.'
NO_TARGET_ELEMENT: Could not find the target element for the color module.