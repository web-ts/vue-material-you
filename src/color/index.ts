import { argbFromHex, themeFromSourceColor, applyTheme } from "@material/material-color-utilities";
import customColors from "./custom-colors.yaml";
import { isDark } from "./dark-mode";
import { isHexColor, rgbFromHex } from "./hex-utilities";

function apply(target: HTMLStyleElement, color: string, dark: boolean) {
  // Create a theme.
  const theme = themeFromSourceColor(argbFromHex(color));
  // Create a dummy element to apply the theme to. This is done due to the limitations of 'applyTheme'.
  const dummy = document.createElement("div");

  // Apply the theme to the dummy element.
  applyTheme(theme, { target: dummy, dark });

  // Get the computed styles of the dummy element and transform all the values to 'r, g, b' strings.
  const data = dummy.style.cssText
    .split(";")
    .map((colorSet) => {
      const [key, value] = colorSet.split(":");

      return value ? { key, value: rgbFromHex(value) } : { key: null, value: null };
    })
    .map(({ key, value }) => (value ? `${key.replace("md-sys", "vmu")}: ${value}` : ""))
    .join(";\n");

  /**
   * Custom Colors
   */
  const cc = Object.entries(customColors[dark ? "dark" : "light"])
    .map(([key, value]) => `${key}: ${value}`)
    .join(";\n");

  target.innerHTML = `* {\n${data}\n${cc}}`;
}

/**
 * Initialize the color module and apply the theme to the body of the document
 * @param color The color that our pallette will be based on
 */
export default function (color: string): void {
  if (!isHexColor(color)) throw new Error("Invalid HEX color. Please use a valid HEX (eg. #12F349) color.");

  const style = document.getElementById("vmu-colors") as HTMLStyleElement ?? document.createElement("style");

  style.id = "vmu-colors";

  apply(style, color, isDark.value);
  document.head.appendChild(style);
  watch(isDark, () => {
    apply(style, color, isDark.value);
  });
}
