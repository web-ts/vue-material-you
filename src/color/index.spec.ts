import { expect, describe, it, afterEach } from "vitest";
import color from ".";
import { isDark, onSchemeChange, setDarkMode, settings } from "./dark-mode";
import { isHexColor, rgbFromHex } from "./hex-utilities";

describe("color", () => {
  afterEach(() => {
    document.head.innerHTML = "";
  });

  it("should create the right color variables", () => {
    color("#6750A4");
    const style = document.head.querySelector("style") as HTMLStyleElement;

    expect(style.innerHTML).toMatchSnapshot();
  });

  it("should not create multiple elements", () => {
    color("#6750A4");
    color("#ffffff");
    const elements = document.head.querySelectorAll("style");

    expect(elements.length).toEqual(1);
  });

  it("should not throw an error for an invalid color", () => {
    try {
      color("invalid");
    } catch (error: any) {
      expect(error.message).toEqual("Invalid HEX color. Please use a valid HEX (eg. #12F349) color.");
    }
  });
});

describe("hex-utilities", () => {
  it("should convert hex to rgb", () => {
    expect(rgbFromHex("#6750a4")).toEqual("103, 80, 164");
    expect(rgbFromHex("#6750A4")).toEqual("103, 80, 164");
    expect(rgbFromHex("#123456")).toEqual("18, 52, 86");
  });

  it("should check if the color is a valid hex color", () => {
    // Good values
    expect(isHexColor("#6750a4")).toBeTruthy();
    expect(isHexColor("#ffffff")).toBeTruthy();

    // Bad values
    expect(isHexColor("6750a4")).toBeFalsy();
    expect(isHexColor("#6750a4a")).toBeFalsy();
    expect(isHexColor("#6750a")).toBeFalsy();
    expect(isHexColor("asdasdasd")).toBeFalsy();
  });
});

describe("dark-mode", () => {
  it("should default to auto", () => {
    expect(settings.userMode).toEqual("auto");
  });

  it("should return the right dark mode state", () => {
    setDarkMode("dark");
    expect(isDark.value).toBeTruthy();
  });

  it("should return the right light mode state", () => {
    setDarkMode("light");
    expect(isDark.value).toBeFalsy();
  });

  it("should return the right system mode", () => {
    onSchemeChange({ matches: false } as MediaQueryListEvent);
    expect(settings.systemMode).toEqual("light");
    onSchemeChange({ matches: true } as MediaQueryListEvent);
    expect(settings.systemMode).toEqual("dark");
  });

  it("should return the right user mode", () => {
    setDarkMode("dark");
    expect(settings.userMode).toEqual("dark");
    setDarkMode("light");
    expect(settings.userMode).toEqual("light");
  });
});
