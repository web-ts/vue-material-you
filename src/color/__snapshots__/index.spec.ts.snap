// Vitest Snapshot v1

exports[`color > should create the right color variables 1`] = `
"* {
--md-sys-color-primary: 103, 80, 164;
 --md-sys-color-on-primary: 255, 255, 255;
 --md-sys-color-primary-container: 233, 221, 255;
 --md-sys-color-on-primary-container: 34, 0, 93;
 --md-sys-color-secondary: 98, 91, 113;
 --md-sys-color-on-secondary: 255, 255, 255;
 --md-sys-color-secondary-container: 232, 222, 248;
 --md-sys-color-on-secondary-container: 30, 25, 43;
 --md-sys-color-tertiary: 126, 82, 96;
 --md-sys-color-on-tertiary: 255, 255, 255;
 --md-sys-color-tertiary-container: 255, 217, 227;
 --md-sys-color-on-tertiary-container: 49, 16, 29;
 --md-sys-color-error: 186, 26, 26;
 --md-sys-color-on-error: 255, 255, 255;
 --md-sys-color-error-container: 255, 218, 214;
 --md-sys-color-on-error-container: 65, 0, 2;
 --md-sys-color-background: 255, 251, 255;
 --md-sys-color-on-background: 28, 27, 30;
 --md-sys-color-surface: 255, 251, 255;
 --md-sys-color-on-surface: 28, 27, 30;
 --md-sys-color-surface-variant: 231, 224, 235;
 --md-sys-color-on-surface-variant: 73, 69, 78;
 --md-sys-color-outline: 122, 117, 127;
 --md-sys-color-outline-variant: 202, 196, 207;
 --md-sys-color-shadow: 0, 0, 0;
 --md-sys-color-scrim: 0, 0, 0;
 --md-sys-color-inverse-surface: 49, 48, 51;
 --md-sys-color-inverse-on-surface: 244, 239, 244;
 --md-sys-color-inverse-primary: 207, 188, 255;

--md-sys-color-disabled: 31, 31, 31}"
`;
